semantics/src/main/scala/aqua/semantics/rules/locations/LocationsState.scala

```diff
@@ -2,25 +2,39 @@ package aqua.semantics.rules.locations
 
 import aqua.parser.lexer.Token
 import aqua.semantics.rules.types.TypesState
+
 import cats.kernel.Monoid
 
 case class LocationsState[S[_]](
-  tokens: List[(String, TokenInfo[S])] = Nil,
-  locations: List[(Token[S], Token[S])] = Nil,
-  stack: List[LocationsState[S]] = Nil
+  tokens: List[(String, ExprInfo[S])] = Nil,
+  locations: List[(Token[S], Token[S])] = Nil
 ) {
 
   lazy val allLocations: List[(Token[S], Token[S])] = locations
+
+  // TODO: optimize distinction
+  def addTokens(newTokens: List[(String, ExprInfo[S])]): LocationsState[S] =
+    copy(tokens = (newTokens ++ tokens).distinct)
+
+  def addToken(newToken: (String, ExprInfo[S])): LocationsState[S] =
+    copy(tokens = (newToken +: tokens).distinct)
+
+  def findTokenByName(
+    name: String,
+    token: Token[S]
+  ): Option[(Token[S], Token[S])] =
+    tokens.find(_._1 == name).map(_._2).map(token -> _.token)
```

Comment:
This function accepts `token` just to return it back? Maybe it is better to refactor?